README
File-Unpack

File::Unpack is an aggressive unpacker for archive files. We call it aggressive, 
because it recursivly descends into any freshly unpacked file, if it appears to be an archive itself.
It also uncompresses files where needed. The ultimate goal of File::Unpack is
to extract as much readable text (ascii or any other encoding) as possible.
Most of the currently known archive file formats are supported.

While The following perl modules are considered helpful, although we may run
perfectly okay, if only a subset of these is available:

 - File::LibMagic, File::MimeInfo::Magic
   File::LibMagic is preferred, its underlying libmagic is activly maintained.
    Its database is very large, and it is much faster than File::MimeInfo::Magic
    Beware, that File::MimeInfo::Magic may derive a mimetype from the suffix, if 
    nothing else works. 
 - IO::Uncompress::AnyUncompress	(perl-IO-Compress)
 - IO::Uncompress::AnyInflate
  - Compress::Raw::Bzip2
  - Compress::Raw::Zlib
 - Archive::Zip
 - String::ShellQuote

Do not use:
 - File::MMagic, File::MMagic::XS, File::Type
   These three are inferior, compared to the other options.
 - Archive::Tar
   It is pure perl, so it's a lot slower then your "/bin/tar"
   IT is heavy on memory, all will be read into memory.
 - Archive::Zip
   If you are just going to be extracting zips (and/or other archives) you are
   recommended to look at using Archive::Extract
 - "Archive::Extract" 
   tries first to determine what type of archive you are passing it, by
   inspecting its suffix. It does not do this by using Mime magic.
   Maybe this module should use something like "File::Type" to determine the
   type, rather than blindly trust the suffix.
   Set $Archive::Extract::PREFER_BIN to 1, which will prefer the use of command
   line programs and won't consume so much memory. Default: use "Archive::Tar" -.




We also rely on a range of command line utilities to be callable via the shell.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc File::Unpack

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=File-Unpack

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/File-Unpack

    CPAN Ratings
        http://cpanratings.perl.org/d/File-Unpack

    Search CPAN
        http://search.cpan.org/dist/File-Unpack/


LICENSE AND COPYRIGHT

Copyright (C) 2010 Juergen Weigert

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
0.69	jw Mon Jul 21 17:21:13 CEST 2014
	- fixed https://github.com/jnweiger/perl-File-Unpack/issues/2

0.67	jw Mon Jul 21 16:14:49 CEST 2014
	- Semantics change of file_unpack.pl:
	  Default to one_shot unless $0 contains 'deep'. New --deep option to also switch this one.
	  This is a move towards least surprise. No other unpacker has this deep feature.

	- Fixed builtin helper for application=x-lzip to conform to IPC::Run::run() semantics.
	- Added an example how to write a mime helper script to the documentation.
	- Fixed diagnostics in case a helper script fails.
	- Several smallish doku improvements.
	- Updated test suite to accept encoding 'unknown' as a synonym to 'binary'.

0.66	jw Mon Jan 27 18:57:16 CET 2014
	- harmful filename test removed, fixing bnc#860561. 
	  The name may not match, when symlinking back. Readlink is the only thing that counts.
	- _run_mime_helper() now returns diagnostics even when 'successful' so that we 
	  can see reasoning for 'passed'.
	- added t/08-survive-corrupt.t to define behavoiur for good10.pdf and bad34.pdf
	- registered the two new test files with t/02-mime.t and t/07-cmdline.t

0.65	jw Tue Aug  6 11:23:43 CEST 2013
	- CPAN no longer fails with  'Using $< in a non-suffix rule context is
	  a GNUmake idiom'

0.64	jw Mon Jul 29 20:54:55 CEST 2013
        - assert writable dirs in _run_mime_helper while disassembling jail.
	  needed for ksh/ast-base.2012-08-01.tar.bz2/src/cmd/pax/data/ro.dat
	- all rename()s in _run_mime_helper() equipped with or die.
	- DESTROY() now handles more cases: Adding the epilog when missing is paramount.

0.63	jw Fri Jun 14 12:08:25 CEST 2013
	- added check for lzip, please test with
	  /space/legaldb_nfs/sha256/pu/fF/6EtJ9m+nlIZ_aDn2nL2QiBahLsFUhJxKcLAv3zA

0.62	jw Tue May  7 20:04:15 CEST 2013
	- support mac pax files. They are gzipped cpio:
	  'ASCII cpio archive (pre-SVR4 or odc)' and are reported by flm
	  as application/octet-stream. 
	  
	(0.61 had a defunct testsuite)

0.60    jw, Mon Apr 29 16:22:33 CEST 2013
        - report text/x-application-xml instead of application/xml

0.59    jw, Tue Mar 26 22:53:51 CET 2013
        - helping upx to survive symlinks. "%{lsrc}s" format added.
        - POD fixes from from anirvan/patch-1
        - application/msword helper added. Using antiword.
	- helper diagnostics improved: first line of stderr is captured in 
	  the log.
	- {verbose} control improved.

0.58	jw, Fri Feb  8 11:15:33 CET 2013
	- bugfix: {inside_archives} was not decremented after leaving an
	  archive.

0.57	jw, Thu Feb  7 19:01:36 CET 2013
	- fixed -f to really allow crawling symlinks.
	  * the logfile will contain the link name, not the link target.
	  * it thus preserves the illusion that all is inside the destination
	    tree.
	- The following command can now be used to recreate the logfile (and
	  retry more unpacking) in $DIR:
	  file_unpack -D $DIR -f -L $DIR.unpack.log $DIR
	- _run_mime_helper(): cleanup $jail_base added when cmd fails to run. 
	  Helper failure may still may result in missing files.
	- t/07-cmdline.t added.

0.56	jw, Tue Feb  5 18:58:06 CET 2013
	- guard _my_shell_quote against undef. 
	- no 'running as root warning' when !{verbose}
	- Relax testsuite on test.mht
	- relax DESTROY to not start a logfile, when none is requested.
	- testsuite really silent now.

0.55	jw, Fri Jan 25 22:09:04 CET 2013
	- option --follow-file-symlinks or new(follow_file_symlinks => $yes) added;
	  where $yes is 0, 1, or 2.
	  We may want to use '1', if we are e.g. inside a legaldb_unpacked 
	  tree, where everything is a symlink. This usage is experimental.
	  A value of '2' is hideous and requires studying perldoc Unpack.pm
	  first.
	- tracking $self->{inside_archives} added. Also experimental.

0.54    jw, Fri Mar  9 17:13:48 CET 2012
        - no longer add special files like fifos, sockets, dev nodes to the logfile.
	- loggable_pathname adds {srcdir => 'input'} if path is in
	  $self->{input}

0.53	jw, Thu Feb 16 22:26:06 UTC 2012
	- new parameter -A and new(archive_name_as_dir=>1) added.
	- new(log_type => 'PLAIN') added, used unless -L

0.52	jw, Wed Feb  1 00:35:47 UTC 2012
	- logfile: input menitoned on 'passed' items.
	- Makefile.PL now actually installs file_unpack (rt#73670).
	  Man page is a hack. Sorry.
	- file_unpack can now unpack multiple archives on the command line.
	- helper/handler wording now consistently speaks of helpers.

0.51  	jw, Wed Jan 25 20:23:22 CET 2012
	- input archives that do not exist in destdir (because they were
	  unpacked)
          now show up in the log prefixed with './input/./' unless fullpath.
	- added loggable_pathname()

0.50    jw, Tue Jan 24 18:43:01 CET 2012
	- made testsuite survive on SLE11_SP1, put IO::Handle::tell() inside
	  eval to survive SLES11_SP1

0.49	jw, Mon Jan 23 19:12:44 CET 2012
	- made survive ntop, in{file} must not modify, so that reopen inside
	  mime() works.

0.48	jw, Tue Nov 29 11:23:30 UTC 2011
	- fixed Use of qw(...) as parentheses. This is deprecated with newer
	  perl.

0.47	jw, Wed Oct  5 00:58:11 CEST 2011
	- fixed rt#70885, new(logfile => \$log); now works as advertised.
	- somewhat experimental code: we now assert there is no utf8 in the log.

0.46	jw, Wed Sep  7 16:08:45 CEST 2011
    	- No more uninitialized value in printf line 780.
        - Incompatible change: unpacked filenames are now relative to destdir
	  unless -F or ->new(log_fullname => 1)
	- New parameter -P or ->new(log_params => {}) to feed params into the 
	  logfile.

0.45	jw, Fri May 27 13:10:16 CEST 2011
	Added option world_readable=>1, to assert 0755/0444 modes for
	dirs/files.

0.44	jw, Fri May 20 12:47:27 CEST 2011
	converted log to use syswrite() attempt to catch write errors...

0.43 	jw, Thu May 12 13:22:02 CEST 2011
	Implemented no_op option in _run_mime_helper().

0.42 	jw, Thu Apr 21 15:56:27 CEST 2011
	Unexpected DESTROY handler closes json logfile properly with an {error}
	message, if any.

0.41	jw, Mon Apr 18 14:55:45 CEST 2011

Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'File::Unpack',
    AUTHOR              => q{Juergen Weigert <jnw@cpan.org>},
    VERSION_FROM        => 'Unpack.pm',
    ABSTRACT_FROM       => 'Unpack.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    BUILD_REQUIRES => {
        'Test::More' => 0,
	'Test::CheckManifest' => '0.9',
	'FindBin' => 0,
	'Test::Pod' => '1.22',
	'Test::Pod::Coverage' => '1.08',
	'Pod::Coverage' => '0.18',
    },
    PL_FILES            => {},
    # Lzma, Bzip2, Zlib are not strictly required. 
    # we need them only for multi-level mimetype recognition.
    PREREQ_PM => {
	'File::LibMagic' => 0,
	'File::MimeInfo::Magic' => 0,
	'Compress::Raw::Lzma' => 0,
	'Compress::Raw::Bzip2' => 0,
	'Compress::Raw::Zlib' => '2.024',
	'BSD::Resource' => 0,
	'Filesys::Statvfs' => 0,

	'Cwd'	=> 0,
	'Carp'	=> 0,
	'File::Path' => 0,
	'File::Temp' => 0,
	'File::Copy' => 0,
	'File::Compare' => 0,
	'JSON' => 0,
	'String::ShellQuote' => 0,
	'IPC::Run' => 0, 
	'Text::Sprintf::Named' => 0,

	'Getopt::Long' => 0,
	'Pod::Usage' => 0,
    },
    # don't use bzip2 for cpan uploads, win32 smoke testers cannot unpack bz2.
    # dist                => { COMPRESS => 'bzip2 -9f', SUFFIX => 'bz2', },
    clean               => { FILES => ['file_unpack', 'File-Unpack-*' ] },
    EXE_FILES => [ 'file_unpack' ],
);

sub MY::postamble
{
  return q{
file_unpack: file_unpack.pl Makefile
	$(CP) $< $@
	## wait, so that -M < -M succeeds in ExtUtils/Command/MM.pm
	(sleep 1; echo .nf; $(PERL) $@ --help) > $(INST_MAN1DIR)/$@.1 || true
};
}



contrib/stringsx.c
/*
 * stringsx.c 
 *
 * (C) 2011 jnw@cpan.org, 
 * Distribute under MIT or any GPL license.
 *
 * A simplified strings tool, similar to the tool that 
 * comes with gnu binutils, but with the following differences
 *
 * - no -e switch. We support all encodings simultaneously
 * - '\0' characters are stripped, and have no effect, unless
 *   multiple '\0' charachters occur in a row.
 * - adjustable fuzzyness: 3 chars in a row with their 8th bit
 *   set are accepted, control chars except '\t', '\n', '\r'
 *   always cut a string. 
 * - Strings need not be '\0' terminated.
 * - no support for file sections. We always scan the entire file.
 *
 * Implemented in both perl and C. Compile the C version, if you 
 * find significant speed issues with the perl version.
 *
 * 2011-11-01, jnw@cpan.org
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>

int
main(int ac, char **av)
{
  int minlen = 10;
  int badcut = 3*1;	// 3 chars of badness 1, or similar

  if (!av[1])
    {
      fprintf(stderr, "Usage: %s file\n", av[0]);
      exit(1);
    }
  FILE *fp = strcmp(av[1], "-") ? fopen(av[1], "r") : stdin;
  if (!fp)
    {
      fprintf(stderr, "%s: %s\n", av[1], strerror(errno));
      exit(1);
    }
  
  int ch;
  int badcount = 0;
  int printing = 0;
  int queuelen = 0;
  int nulseen = 0;
  char queuebuf[20];

  while ((ch = getc(fp)) != EOF)
    {
      int badness = 0;

      if (ch == 0)
        { 
	  nulseen++;	// a nul every second char is just fine.
	  if (nulseen > 1) badness = badcut+1;
	}
      else
        {
	  nulseen = 0;
	  if (ch > 127)				badness = 1;	// latin1 or utf8 byte
          else if (ch < 32 && ch != '\t' && 
	           ch != '\n' && ch != '\r') 	badness = badcut+1;	// control char.
          else /* (good char) */		badness = 0;
	}

      badcount += badness;

      if (!printing && !badness)
        {
	  queuebuf[queuelen] = ch;
	  if (ch) queuelen++;	// always skip \0 bytes
	  if (queuelen >= minlen) 
	    {
	      int j;
	      for (j = 0; j < queuelen; j++) 
	        putchar(queuebuf[j]);
	      queuelen = 0;
	      printing = 1;
	    }
	  continue;
	}

      if (printing)
        {
	  if (!badness && ch)
	    {
	      if (queuelen)
	        {
	          int j;
	          for (j = 0; j < queuelen; j++) 
	            putchar(queuebuf[j]);
	          queuelen = 0;
		}
	      queuelen = 0;
	      badcount = 0;
	      putchar(ch);
	    }
	  else
	    {
	      queuebuf[queuelen] = ch;
	      if (ch) queuelen++;	// always skip \0 bytes
	      if (badcount >= badcut) 
	        { 
		  queuelen = 0;
	          printing = 0;
		  badcount = 0;
		  putchar('\f');	// next string.
		  putchar('\n');	// next string.
		}
	    }
	}
    }
  fclose(fp);
  return 0;
}


contrib/stringsx.pl
#! /usr/bin/perl -w
#
# stringsx.pl
#
# (C) 2011 jnw@cpan.org
# Distribute under MIT or any GPL license.
#
# A simplified strings tool, similar to the tool that 
# comes with gnu binutils, but with the following differences
#
# * no -e switch. We support all encodings simultaneously
# * '\0' characters are stripped, and have no effect, unless
#   multiple '\0' charachters occur in a row.
# * adjustable fuzzyness: 3 chars in a row with their 8th bit
#   set are accepted, control chars except '\t', '\n', '\r'
#   always cut a string. 
# * Strings need not be '\0' terminated.
# * no support for file sections. We always scan the entire file.
#
# Implemented in both perl and C. Compile the C version, if you 
# find significant speed issues with the perl version.
#
# 2011-11-01, jnw@cpan.org

my $infile = shift;
unless (defined $infile)
  {
    die "Usage: $0 file\n";
  }

my $fd = STDIN;
if ($infile ne '-')
  {
    open $fd, "<", $infile or die "open $infile failed: $!\n";
  }

my $minlen = 10;
my $badcut = 3*1;	# 3 chars of badness 1, or similar

my $ch;
my $badcount = 0;
my $printing = 0;
my $queuelen = 0;
my $nulseen = 0;
my $queuebuf = '';

while (defined($ch = getc($fd)))
  {
    my $badness = 0;
    my $oc = ord($ch);
    if ($oc == 0)
      {
 	$nulseen++;	# a nul every second char is just fine.
 	if ($nulseen > 1) { $badness = $badcut+1; }
      }
    else
      {
 	$nulseen = 0;
 	if ($oc > 127)                     { $badness = 1; }	     # latin1 or utf8 byte
	elsif ($oc < 32 && $ch ne "\t" && 
 	       $ch ne "\n" && $ch ne "\r") { $badness = $badcut+1; } #  control char.
        else 				   { $badness = 0; }	     # good char
      }

# 
    $badcount += $badness;

    if (!$printing && !$badness)
      {
        $queuebuf .= $ch if $oc; 	# always skip \0 bytes

        if (length($queuebuf) >= $minlen) 
          {
	    print $queuebuf;
            $queuebuf = '';
            $printing = 1;
          }
        next;
      }

    if ($printing)
      {
        if (!$badness && $oc)
          {
            if (length($queuebuf))
              {
	        print $queuebuf;
		$queuebuf = '';
      	      }
            $badcount = 0;
            print $ch;
          }
        else
          {
            $queuebuf .= $ch if $oc; 	# always skip \0 bytes
            if ($badcount >= $badcut) 
              { 
      	        $queuebuf = '';
                $printing = 0;
      	  	$badcount = 0;
      	  	print "\f\n";		# next string.
      	      }
          }
      }
  }
close $fd;



file_unpack.pl
#! /usr/bin/perl -w
#
# file_unpack.pl -- Demo of File::Unpack features.
# 
# (C) 2010-2014, jnw@cpan.org, all rights reserved.
# Distribute under the same license as Perl itself.
#
# 2010-06-29, jw -- initial draught
# 2010-08-03, jw -- fixed -v.
# 2010-08-31, jw -- fixed -q with -m.
# 2010-09-01, jw -- added --list
# 2011-01-03, jw -- allow multiple arguments. Improved -m
# 2011-03-08, jw -- fixed usage of -l, added -p.
# 2011-04-21, jw -- better format error messages, and stop after error.
# 2011-05-12, jw -- added -n for no_op
# 2012-02-16, jw -- added -A for archive_name_as_dir.
#                   using {log_type} == 'PLAIN' unless -L
# 2013-01-25, jw -- added -f for follow_file_symlinks.
# 2014-07-21, jw -- default to one_shot unless named *deep*.

use Data::Dumper;
$Data::Dumper::Terse = 1;
$Data::Dumper::Indent = 1;
use Getopt::Long qw(:config no_ignore_case);
use Pod::Usage;
use FindBin;
use lib "$FindBin::RealBin/blib/lib";
use File::Unpack;

my $version = $File::Unpack::VERSION;
my @exclude;
my $exclude_vcs = 1;
my $help;
my $mime_only;
my $list_only;
my $list_perlish;
my @mime_helper_dirs;

my %opt = ( verbose => 1, maxfilesize => '2.6G', one_shot => 1, no_op => 0, world_readable => 0, log_fullpath => 0, archive_name_as_dir => 0, follow_file_symlinks => 0, log_type => 'PLAIN');

$opt{one_shot} = 0 if $0 =~ m{deep[^/]*$};

push @mime_helper_dirs, "$FindBin::RealBin/helper" if -d "$FindBin::RealBin/helper";

GetOptions(
	"verbose|v+"   	=> \$opt{verbose},
	"version|V"    	=> sub { print "$version\n"; exit },
	"quiet"        	=> sub { $opt{verbose} = 0; },
	"destdir|D|C=s" => \$opt{destdir},
	"exclude|E=s"  	=> \@exclude,
	"exclude-vcs!" 	=> \$exclude_vcs,
	"vcs|include-vcs!" 	=> sub { $exclude_vcs = !$_[1]; },
	"help|?"       		=> \$help,
	"logfile|L=s"  		=> \$opt{logfile},
	"fullpath-log|F" 	=> \$opt{log_fullpath},
	"one-shot|one_shot|1"   => \$opt{one_shot},
	"deep|recursive"        => sub { $opt{one_shot} = 0; },
	"mimetype|m+"  		=> \$mime_only,
	"no_op|no-op|noop|n+" 	=> \$opt{no_op},
	"list-helpers|l+" 	=> \$list_only,
	"print-helpers|p+" 	=> \$list_perlish,
	"params|P=s"		=> \%{$opt{log_params}},
	"maxfilesize=s"		=> \$opt{maxfilesize},
	"use-mime-helper-dir|I|u=s" 		=> \@mime_helper_dirs,
	"world-readable|world_readable|R+" 	=> \$opt{world_readable},
	"archive-dirs|archive_dirs|A"		=> \$opt{archive_name_as_dir},
	"follow-file-symlinks|f+" 		=> \$opt{follow_file_symlinks},
) or $help++;

@mime_helper_dirs = split(/,/,join(',',@mime_helper_dirs));
my $archive = shift or $list_perlish or $list_only or $help++;

pod2usage(-verbose => 1, -msg => qq{
file_unpack V$version Usage: 

$0 [options] input.tar.gz ...
$0 [options] input/ ...
$0 -l
$0 -p

Valid options are:
 -v	Be more verbose. Default: $opt{verbose}.
 -q     Be quiet, not verbose.

 -A --archive_dirs
	Use exact archive names as subdirectories, modified by appending '._*'
	to avoid collisions.  Default: use truncated and/or modified archive
	names. E.g. example.zip is per default unpacked into 'example/'. With
	-A, it is unpacked into 'example.zip._/'.  This does not apply for
	single file archives. They are always unpacked without a directory.

 -C dir
 -D --destdir dir
        Directory, where to place the output file or directory.
	A subdirectory is created, if there are more than one files to unpack.
	Default: current dir.

 -E --exclude glob.pat
 	Specify files and directories that are not unpacked.
	This option can be specified multiple times.

 -F --fullpath-log
 	Always use full path names in logfile. Default: 
	unpacked path names are written relative to destdir.

 --exclude-vcs	--no-exclude-vcs 
 --include-vcs  --no-include-vcs --vcs --no-vcs
 	Group switch for directory glob patterns of most version control systems.
	This affects at least SCCS, RCS, CVS, .svn, .git, .hg, .osc .
	The logfile has a {skipped}{exclude} counter.
        Default: exclude-vcs=$exclude_vcs .

 -1 --one-shot
 	Make unpacker non-recursive. Perform one level of unpacking only.
	This is the default unless the name of the unpacker contains the substring 'deep'.

 --deep
 --recursive
 	Make unpacker recursive. Perform all possible levels of unpacking.
	This is the default if the name of the unpacker contains the substring 'deep'.

 -h --help -?
        Print this online help.
 
 -L --logfile  file.log
 	Specify a logfile, where freshly unpacked files are reported.
	When a logfile is specified, its format is JSON; 
	default is STDOUT with format PLAIN.
 
 -l --list-helpers
        Overview of mime-type patterns and their helper commands.

 -p --print-helpers
 	List all builtin mime-helpers and all external mime-helpers as 
	a nested perl datastructure.

 -P --param KEY=VALUE
 	Place additional params into the log file.

 --maxfilesize size
        Truncate an unpacked file, if it gets larger than the specified size.
	Size can be specified as bytes (plain integer), kilo-, mega-, giga-, or 
	tera-bytes (suffix K,M,G,T). Default: $opt{maxfilesize}.

 -m --mimetype
        Do not unpack, just report mimetype of the archive. Output format is 
	similar to '/usr/bin/file -i', unless -q or -v are given.
	With -v, the unpacker command is also printed.

 -R --world-readable
 	Make the unpacked tree world readable. Default: user readable.

 -n --no-op
 	Do not unpack. Print the first unpack command only.

 -u --use-mime-helper-dir dir
 	Include an additonal directory of mime helpers.
	Useable multiple times. Later additions take precedence.

 -f --follow-file-symlinks
 	Follow (and unpack) symlinks that point to files (or archives).
	Used once, we follow only symlinks that are present before unpacking starts.
	Used twice, we also follow symlinks that were unpacked from an archive.
	Symlinks to directories or other (dangling) symlinks are always ignored.
	The logfile has a {skipped}{symlink} counter.
        Default: skip all symlinks.

}) if $help;

if (defined $opt{logfile})
  {
    $opt{log_type} = 'JSON';
    $opt{logfile} = \*STDOUT if $opt{logfile} eq '-';
  }

$opt{logfile} ||= '/dev/null' if $list_only or $list_perlish or $mime_only or $opt{no_op};
my $u = File::Unpack->new(%opt);
my $list = $u->mime_helper_dir(@mime_helper_dirs);

if ($list_perlish)
  {
    print Dumper $list;
    exit 0;
  }

if ($list_only)
  {
    printf @$_ for $u->list();
    exit 0;
  }

if ($mime_only)
  {
    $u->{verbose}-- if $u->{verbose};
    if ($u->{verbose} > 1)
      {
        print "using File::LibMagic $File::LibMagic::VERSION\n" if defined $File::LibMagic::VERSION;
        print "using File::MimeInfo::Magic $File::MimeInfo::Magic::VERSION\n" if defined $File::MimeInfo::Magic::VERSION;
        print "using File::Unpack $File::Unpack::VERSION\n" if defined $File::Unpack::VERSION;
      }

    while (defined $archive)
      {
	my $m = $u->mime($archive);
	my ($h,$r) = $u->find_mime_helper($m);
	if ($opt{verbose} > 1)
	  {
	    print "$archive: ", Dumper $m;
	    print File::Unpack::fmt_run_shellcmd($h) . "\n";
	  }
	elsif ($opt{verbose} == 1)
	  {
	    print "$archive: $m->[0]; charset=$m->[1]\n";
	  }
	else
	  {
	    print "$m->[0]\n";
	  }
        $archive = shift;
      }
    exit 0;
  }

while (defined $archive and !$u->{error}) 
  {
    $u->exclude(vcs => $exclude_vcs);
    $u->exclude(add => \@exclude) if @exclude;

    $u->unpack($archive);
    map { print STDERR "ERROR: $_\n" } @{$u->{error}} if $u->{error};
    $archive = shift;
    if (defined($archive))
      {
        if (defined $opt{logfile} and -f $opt{logfile})
	  {
            warn "File::Unpack($archive): overwriting previous logfile $opt{logfile} in 3 seconds. Press CTRL-C to abort.\n" if defined $archive;
	    sleep(3);
	  }
        # reload, for the next round. (new() opens the logfile, unpack() closese it.)
        $u = File::Unpack->new(%opt);
        $u->mime_helper_dir(@mime_helper_dirs);
      }
  }

# delete $u->{json};
# die "$0: " . Dumper $u;
